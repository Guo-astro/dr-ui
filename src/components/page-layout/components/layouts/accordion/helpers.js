// frontmatter (require) object. This is the page's generated frontMatter.
// headings (optional) array. This is used by api-documentation to combine multiple markdown files.

export function buildSecondLevel(frontMatter, headings) {
  if (!frontMatter.headings && !headings) return [];
  // parse headings
  const orderedHeadings = frontMatter.headings
    ? parseHeadings(frontMatter, frontMatter.headings)
    : parseHeadings(frontMatter, headings);
  // find top level headings (h2)
  const topLevelHeadings = orderedHeadings.filter((h) => h.level === 2);
  // return data
  return buildItems(topLevelHeadings, orderedHeadings);
}

function buildItems(topLevelHeadings, orderedHeadings) {
  return topLevelHeadings.map((heading, index) => {
    const nextHeading = topLevelHeadings[index + 1];
    return {
      title: heading.text,
      path: heading.slug,
      ...(heading.tag && { tag: heading.tag }),
      thirdLevelItems: buildThirdLevel(orderedHeadings, nextHeading, heading)
    };
  });
}

function buildThirdLevel(orderedHeadings, nextHeading, h2) {
  const items = orderedHeadings.filter(
    (f) =>
      f.level === 3 &&
      f.order > h2.order &&
      (nextHeading ? f.order < nextHeading.order : true)
  );
  return items.map((heading) => {
    return {
      title: heading.text,
      path: heading.slug,
      ...(heading.tag && { tag: heading.tag || undefined })
    };
  });
}

function parseHeadings(frontMatter, headings) {
  return headings.map((heading, index) => {
    const tag = parseBeta(frontMatter, heading);
    return {
      level: heading.level,
      text: heading.text,
      slug: heading.slug,
      order: index,
      ...(tag && { tag: tag })
    };
  });
}

function parseBeta(frontMatter, heading) {
  return frontMatter.beta && frontMatter.beta.indexOf(heading.text) > -1
    ? 'beta'
    : undefined;
}
