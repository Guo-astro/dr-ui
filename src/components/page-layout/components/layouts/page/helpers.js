import React from 'react';
import classnames from 'classnames';

export function buildSections(headings) {
  if (!headings) return [];
  return headings
    .filter((heading) => heading.level === 2 || heading.level === 3)
    .map((heading) => ({
      title: heading.text,
      url: `#${heading.slug}`,
      level: heading.level
    }));
}

export function buildScollspyItems(sections) {
  return sections.map((heading) => heading.url.replace('#', ''));
}

export function buildToc(sections) {
  return sections.map((heading, index) => {
    const nextHeading = sections[index + 1] || null;
    const headingClasses = classnames('', {
      pb6: heading.level === 2,
      'mb6 color-gray': heading.level === 3,
      mb24: nextHeading && nextHeading.level === 2 && heading.level === 3,
      mb6: nextHeading && nextHeading.level === 2 && heading.level === 2
    });
    return (
      <li key={index} className={headingClasses}>
        <a href={heading.url} className="color-blue-on-hover">
          {heading.title}
        </a>
      </li>
    );
  });
}
