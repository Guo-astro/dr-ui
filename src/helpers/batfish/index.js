const { buildNavigation } = require('./navigation');
const { buildTopics } = require('./topics');
const { buildSplitPages } = require('./split-pages');

module.exports = {
  buildNavigation,
  buildTopics,
  buildSplitPages
};
