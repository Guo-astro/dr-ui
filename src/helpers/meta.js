// this function returns `meta` for ReactPageShell
// fields to return if data is available
const fields = [
  'title',
  'description',
  'pathname',
  'contentType',
  'language',
  'level',
  'hideFromSearchEngines'
];
// metamaker accepts three objects: frontmatter, location, meta (optional)
// these props come from PageShell
export function metaMaker(frontMatter, location, meta) {
  return fields.reduce((obj, field) => {
    // check if the field exists in the (existing) meta, frontMatter, or location objectys
    const value =
      (meta ? meta[field] : null) ||
      (frontMatter ? frontMatter[field] : null) ||
      (location ? location[field] : null);
    // if the field exists, push the field and value to the meta object
    if (value) obj[field] = value;
    return obj;
  }, {});
}
