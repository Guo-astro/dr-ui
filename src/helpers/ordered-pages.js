// builds firstLevelItems for NavigationAccordion
// should be used in batfish.config.js `dataSelectors`:
// orderedPages: data => buildOrderedPages(data, productPageOrder)
// productPageOrder = object of paths
export function buildOrderedPages(data, productPageOrder) {
  const pages = data.pages.map(p => ({
    title: p.frontMatter.title,
    path: p.path,
    tag: p.frontMatter.tag || '',
    description: p.frontMatter.description
  }));

  return Object.keys(productPageOrder).reduce((reduced, order) => {
    const items = productPageOrder[order];
    reduced[order] = items.map(item => {
      return pages.find(p => {
        const urlEnding = item === 'index' ? '' : `${item}/`;
        return new RegExp(`${order}${urlEnding}$`).test(p.path);
      });
    });
    return reduced;
  }, {});
}
