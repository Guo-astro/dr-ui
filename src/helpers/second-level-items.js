// builds secondLevelItems for NavigationAccordion
// `headings` = this.props.frontMatter.headings || this.props.headings
export function buildSecondLevelItems(headings) {
  const orderedHeadings = headings.map((heading, index) => {
    return {
      level: heading.level,
      text: heading.text,
      slug: heading.slug,
      order: index
    };
  });
  const topLevelHeadings = orderedHeadings.filter(h => h.level === 2);
  return topLevelHeadings.map((h2, index) => {
    const nextH2 = topLevelHeadings[index + 1];
    return {
      title: h2.text,
      path: h2.slug,
      thirdLevelItems: orderedHeadings
        .filter(
          f =>
            f.level === 3 &&
            f.order > h2.order &&
            (nextH2 ? f.order < nextH2.order : true)
        )
        .map(h3 => {
          return {
            title: h3.text.replace(/i2157|i2860/, ''),
            path: h3.slug
          };
        })
    };
  });
}
