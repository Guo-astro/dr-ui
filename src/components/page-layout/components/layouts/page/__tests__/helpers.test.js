import { buildScollspyItems, buildSections, buildToc } from '../helpers.js';

it('buildSections', () => {
  expect(
    buildSections([
      { text: 'Heading 1', slug: 'heading-1', level: 2 },
      { text: 'Heading a', slug: 'heading-a', level: 3 },
      { text: 'Heading b', slug: 'heading-b', level: 3 },
      { text: 'Heading c', slug: 'heading-c', level: 3 },
      { text: 'Heading 2', slug: 'heading-2', level: 2 },
      { text: 'Heading a', slug: 'heading-a-1', level: 3 },
      { text: 'Heading b', slug: 'heading-b-1', level: 3 },
      { text: 'Heading c', slug: 'heading-c-1', level: 3 }
    ])
  ).toEqual([
    { title: 'Heading 1', url: '#heading-1', level: 2 },
    { title: 'Heading a', url: '#heading-a', level: 3 },
    { title: 'Heading b', url: '#heading-b', level: 3 },
    { title: 'Heading c', url: '#heading-c', level: 3 },
    { title: 'Heading 2', url: '#heading-2', level: 2 },
    { title: 'Heading a', url: '#heading-a-1', level: 3 },
    { title: 'Heading b', url: '#heading-b-1', level: 3 },
    { title: 'Heading c', url: '#heading-c-1', level: 3 }
  ]);
});

it('buildToc', () => {
  expect(
    buildToc([
      { title: 'Heading 1', url: '#heading-1', level: 2 },
      { title: 'Heading a', url: '#heading-a', level: 3 },
      { title: 'Heading b', url: '#heading-b', level: 3 },
      { title: 'Heading c', url: '#heading-c', level: 3 },
      { title: 'Heading 2', url: '#heading-2', level: 2 },
      { title: 'Heading a', url: '#heading-a-1', level: 3 },
      { title: 'Heading b', url: '#heading-b-1', level: 3 },
      { title: 'Heading c', url: '#heading-c-1', level: 3 }
    ])
  ).toMatchSnapshot();
});

it('buildScollspyItems', () => {
  expect(
    buildScollspyItems([
      { title: 'Heading 1', url: '#heading-1', level: 2 },
      { title: 'Heading a', url: '#heading-a', level: 3 },
      { title: 'Heading b', url: '#heading-b', level: 3 },
      { title: 'Heading c', url: '#heading-c', level: 3 },
      { title: 'Heading 2', url: '#heading-2', level: 2 },
      { title: 'Heading a', url: '#heading-a-1', level: 3 },
      { title: 'Heading b', url: '#heading-b-1', level: 3 },
      { title: 'Heading c', url: '#heading-c-1', level: 3 }
    ])
  ).toEqual([
    'heading-1',
    'heading-a',
    'heading-b',
    'heading-c',
    'heading-2',
    'heading-a-1',
    'heading-b-1',
    'heading-c-1'
  ]);
});
