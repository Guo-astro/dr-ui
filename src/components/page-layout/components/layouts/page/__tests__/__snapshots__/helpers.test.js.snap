// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildToc 1`] = `
Array [
  <li
    className="pb6"
  >
    <a
      className="color-blue-on-hover"
      href="#heading-1"
    >
      Heading 1
    </a>
  </li>,
  <li
    className="mb6 color-gray"
  >
    <a
      className="color-blue-on-hover"
      href="#heading-a"
    >
      Heading a
    </a>
  </li>,
  <li
    className="mb6 color-gray"
  >
    <a
      className="color-blue-on-hover"
      href="#heading-b"
    >
      Heading b
    </a>
  </li>,
  <li
    className="mb6 color-gray mb24"
  >
    <a
      className="color-blue-on-hover"
      href="#heading-c"
    >
      Heading c
    </a>
  </li>,
  <li
    className="pb6"
  >
    <a
      className="color-blue-on-hover"
      href="#heading-2"
    >
      Heading 2
    </a>
  </li>,
  <li
    className="mb6 color-gray"
  >
    <a
      className="color-blue-on-hover"
      href="#heading-a-1"
    >
      Heading a
    </a>
  </li>,
  <li
    className="mb6 color-gray"
  >
    <a
      className="color-blue-on-hover"
      href="#heading-b-1"
    >
      Heading b
    </a>
  </li>,
  <li
    className="mb6 color-gray"
  >
    <a
      className="color-blue-on-hover"
      href="#heading-c-1"
    >
      Heading c
    </a>
  </li>,
]
`;
